File: leetcode/p00739/p00739.go
package p00739

// Time Complexity: O(n)
// Space Complexity: O(n)
func dailyTemperatures(temperatures []int) []int {
	res := make([]int, len(temperatures))

	stack := CreateStack()
	for cur, temp := range temperatures {
		for !stack.IsEmpty() && temp > temperatures[stack.Peek()] {
			res[stack.Peek()] = cur - stack.Peek()
			stack.Pop()
		}
		stack.Put(cur)
	}
	return res
}

type Stack struct {
	arr  []int
	size int
}

func CreateStack() Stack {
	return Stack{
		arr:  make([]int, 0),
		size: 0,
	}
}

func (s *Stack) IsEmpty() bool {
	return s.size == 0
}

func (s *Stack) Put(val int) {
	s.arr = append(s.arr, val)
	s.size++
}

func (s *Stack) Pop() {
	if s.IsEmpty() {
		return
	}

	s.arr = s.arr[:s.size-1]
	s.size--
}

func (s *Stack) Peek() int {
	return s.arr[s.size-1]
}

File: leetcode/p00739/p00739.py
from typing import List

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        res = [0] * len(temperatures)
        stack = []

        for cur, temp in enumerate(temperatures):
            # using stack[len(stack)-1] instead of stack[-1]
            while len(stack) > 0 and temp > temperatures[stack[len(stack)-1]]:
                prev = stack.pop()
                res[prev] = cur - prev
            
            stack.append(cur)
        
        return res

if __name__ == '__main__':
    s = Solution()

    assert s.dailyTemperatures([73,74,75,71,69,72,76,73]) == [1,1,4,2,1,1,0,0]
    assert s.dailyTemperatures([30,40,50,60]) == [1,1,1,0]
    assert s.dailyTemperatures([30,60,90]) == [1,1,0]